app_check/integration_test/src/integration_test.cc

```diff
@@ -821,4 +919,191 @@ TEST_F(FirebaseAppCheckTest, TestFunctionsFailure) {
                     firebase::functions::kErrorUnauthenticated);
 }
 
+TEST_F(FirebaseAppCheckTest, TestFirestoreSetGet) {
+  InitializeAppCheckWithDebug();
+  InitializeApp();
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"str", firebase::firestore::FieldValue::String("foo")},
+          {"int", firebase::firestore::FieldValue::Integer(123)}}),
+      "document.Set");
+  firebase::Future<firebase::firestore::DocumentSnapshot> future =
+      document.Get(firebase::firestore::Source::kServer);
+  WaitForCompletion(future, "document.Get");
+  EXPECT_NE(future.result(), nullptr);
+  EXPECT_THAT(future.result()->GetData(),
+              UnorderedElementsAre(
+                  Pair("str", firebase::firestore::FieldValue::String("foo")),
+                  Pair("int", firebase::firestore::FieldValue::Integer(123))));
+}
+
+TEST_F(FirebaseAppCheckTest, TestFirestoreSetGetFailure) {
+  // Don't set up AppCheck
+  InitializeApp();
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  // Both operations should fail because AppCheck isn't configured.
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"str", firebase::firestore::FieldValue::String("badfoo")},
+          {"int", firebase::firestore::FieldValue::Integer(456)}}),
+      "document.Set", firebase::firestore::kErrorPermissionDenied);
+  WaitForCompletion(document.Get(firebase::firestore::Source::kServer),
+                    "document.Get",
+                    firebase::firestore::kErrorPermissionDenied);
+
+  CleanupFirestore(firebase::firestore::kErrorPermissionDenied);
+}
+
+TEST_F(FirebaseAppCheckTest, TestFirestoreTransaction) {
+  InitializeAppCheckWithDebug();
+  InitializeApp();
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("start")}}),
+      "document.Set 0");
+
+  std::vector<firebase::firestore::MapFieldValue> document_snapshots;
```

Comment:
Optional: Technically, there is a race condition accessing `document_snapshots` since it is written to by one thread and read from another. If this test ever flakes, consider protecting the vector with a mutex. Also, if this test case were to uncover bugs it could also crash, such as if a callback to the snapshot listener occurred after the vector was destructed.

Although it adds some noise to the test case, I've fixed this in other tests by creating a struct that contains both the mutex and the vector, putting that struct into a shared_ptr, and capturing that shared_ptr by value in the lambda. Doing this guarantees that the vector and the mutex stay alive as long as they are referenced.

e.g.

```cpp
struct ListenerSnapshots {
  std::mutex mutex;
  std::vector<firebase::firestore::MapFieldValue> snapshots;
};
auto listener_snapshots = std::make_shared<ListenerSnapshots>();
```

Reply:
Added it, just to be safe.  This test is something I copied and simplified from the other Firestore testapps.  I don't think the flakes we are seeing in those are these functions, but if they are we might want to consider updating them to be a bit safer with this logic.